dds_web/api/user.py

```diff
@@ -196,46 +196,52 @@ def invite_user(email, new_user_role, project=None, unit=None):
-                            from_user=auth.current_user(),
-                            to_another=new_invite,
-                            from_user_token=dds_web.security.auth.obtain_current_encrypted_token(),
-                            project=unit_project,
-                        )
-
-                if not project:  # specified project is disregarded for unituser invites
-                    msg = f"{str(new_invite)} was successful."
-                else:
-                    msg = f"{str(new_invite)} was successful, but specification for {str(project)} dropped. Unit Users have automatic access to projects of their unit."
+        try:
+            # Append invite to unit if applicable
+            if new_invite.role in ["Unit Admin", "Unit Personnel"]:
+                # TODO Change / move this later. This is just so that we can add an initial Unit Admin.
+                if auth.current_user().role == "Super Admin":
+                    if unit:
+                        unit_row = models.Unit.query.filter_by(public_id=unit).one_or_none()
+                        if not unit_row:
+                            raise ddserr.DDSArgumentError(message="Invalid unit publid id.")
+
+                        unit_row.invites.append(new_invite)
+                    else:
+                        raise ddserr.DDSArgumentError(message="Cannot invite this user.")
+
+                if "Unit" in auth.current_user().role:
+                    # Give new unit user access to all projects of the unit
+                    auth.current_user().unit.invites.append(new_invite)
+                    for unit_project in auth.current_user().unit.projects:
+                        if unit_project.is_active:
+                            share_project_private_key(
+                                from_user=auth.current_user(),
+                                to_another=new_invite,
+                                from_user_token=dds_web.security.auth.obtain_current_encrypted_token(),
+                                project=unit_project,
+                            )
+
+                    if not project:  # specified project is disregarded for unituser invites
+                        msg = f"{str(new_invite)} was successful."
+                    else:
+                        msg = f"{str(new_invite)} was successful, but specification for {str(project)} dropped. Unit Users have automatic access to projects of their unit."
 
-        else:
-            db.session.add(new_invite)
-            if project:
-                share_project_private_key(
-                    from_user=auth.current_user(),
-                    to_another=new_invite,
-                    project=project,
-                    from_user_token=dds_web.security.auth.obtain_current_encrypted_token(),
-                    is_project_owner=new_user_role == "Project Owner",
-                )
+            else:
+                db.session.add(new_invite)
```

Comment:
Mind that adding the invite to the Database session now happens inside your `try` statement. If a `KeyNotFoundError` is raised, the invite will not be stored in the database (but the mail was already send before)